daemon/algod/api/server/auth.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package server
+
+import (
+	"crypto/subtle"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/labstack/echo/v4"
+
+	"github.com/algorand/go-algorand/daemon/algod/api/server/lib/middlewares"
+	"github.com/algorand/go-algorand/logging"
+)
+
+const urlAuthFormatter = "/urlAuth/%s"
+const debugRouteName = "debug"
+
+// AuthRoutes define the desired mapping of authentication token keys to allowed routes.
+type AuthRoutes map[string]map[echo.Route]echo.HandlerFunc
+
+// Auth is the authentication layer manager. It creates a wrapper around each invoked call and provide authentication.
+type Auth struct {
+	log  logging.Logger
+	echo *echo.Echo
+}
+
+// MakeAuth create an authentication object
+func MakeAuth(log logging.Logger, e *echo.Echo) *Auth {
+	return &Auth{
+		log:  log,
+		echo: e,
+	}
+}
+
+// RegisterHandlers registers the given authentication routes
+func (auth *Auth) RegisterHandlers(authRoutes AuthRoutes) {
+	routeTokens := make(map[echo.Route][]string)
+
+	// for each unique route, find the list of tokens.
+	for token, routes := range authRoutes {
+		for route := range routes {
+			routeTokens[route] = append(routeTokens[route], token)
+		}
+	}
```

Comment:
seems like it would simplify this code as well.

Reply:
it's not incorrect, but from a design point of view, it seems weird to store every token on every route it can authenticate. It seems like there should be something that maps `token -> token type` and then something that maps `token type -> allowed routes`

Reply:
(or `route -> allowed token types` as it may be)

Reply:
that's a nice optimization, for above.. I don't see it becoming a critical unless we're going to have either large number of tokens or large number of entry points.